### PR TITLE
Make `Address:p2sh_from_hash` public

### PR DESCRIPTION
**Pull request by tcharding**
We previously made this function Private and added a comment that doing so was somehow better to remove the footgun of hashing the wrong length script. However in hindsight this was a bad idea and users want the functionality.
    
Make the `Address:p2sh_from_hash` public and document it as we do for `Address::p2sh`.

This is an additive change and is expected to be backported to `v0.32`, as part of the fix to #2784. Please note it introduces the footgun that is described in the function rustdoc. This will be improved as a separate patch and added to the current release.

### CODE REVIEW

**Comment by tcharding** [comment]
Just changed the commit log for the last patch, we are going to end up with millions of commits that all just run `just check-api` and commit the changes to the `api/` files. Since I just mentioned to @jamillambert that typically we do not want to use the same commit message in multiple places too much I thought I'd mention it.

Are you ok with me using `git commit -a -m 'api: Run just check-api'` for all of these patches @apoelstra?

**Review by sanket1729 — DISMISSED** [review]
ACK 3a2e11985a7ba6cc402627410b767ca91a09dc44

**Comment by tcharding** [comment]
@sanket1729! Thanks for the review man.

File: bitcoin/src/address/mod.rs [hunk 1]
```diff
@@ -384,8 +384,11 @@ impl Address {
         Ok(Address::p2sh_from_hash(hash, network))
     }
 
-    // This is intentionally not public so we enforce script length checks.
-    fn p2sh_from_hash(hash: ScriptHash, network: impl Into<NetworkKind>) -> Address {
+    /// Creates a pay to script hash P2SH address from a script hash.
+    ///
+    /// **Warning** The `hash` pre-image (redeem script) must not exceed 520 bytes in length
```
Comment (storopoli) [inline bitcoin/src/address/mod.rs:389]: I think we should use the standard proposed in #2798:

```suggestion
    /// # Warning
    ///
    /// The `hash` pre-image (redeem script) must not exceed 520 bytes in length
```

Comment (tcharding) [reply bitcoin/src/address/mod.rs:389]: Oooo, good point. Thanks man, will do.

Comment (tcharding) [reply bitcoin/src/address/mod.rs:389]: Maybe plural to fit in with `Examples` and `Returns`?

**Comment by tcharding** [comment]
Changes in force push:

- Use `# Warning` header
- Put the docs change in the right patch instead of in the 'just check-api` patch
- Do five push ups for not reading my diff before pushing last time ( cc @storopoli :)

**Review by apoelstra — APPROVED** [review]
ACK 3615410d215222ca5fdb431706007a36df5c797d

**Comment by apoelstra** [comment]
Needs backport.

**Comment by tcharding** [comment]
Backport done: https://github.com/rust-bitcoin/rust-bitcoin/pull/2806

**Pull request merged**